trex-scripts/tests/qos_strict_priority.py

```diff
@@ -0,0 +1,227 @@
+        # Create the system stream
+        system_pg_id = 2
+        system_stream = self.create_system_stream(system_pg_id)
+        self.client.add_streams(system_stream, ports=PRIORITY_SENDER_PORT)
+
+        # Start sending traffic
+        logging.info(
+            "Starting traffic, duration: %d sec", args.duration,
+        )
+        self.client.start(ALL_SENDER_PORTS, mult='1', duration=args.duration)
+
+        logging.info("Waiting until all traffic is sent")
+        self.client.wait_on_traffic(ports=ALL_SENDER_PORTS, rx_delay_ms=100)
+
+        # Get latency stats
+        stats = self.client.get_stats()
+        lat_stats = stats['latency'].get(pg_id)
+        flow_stats = stats["flow_stats"].get(pg_id)
+        pp.pprint(stats)
+
+        print("Flow stats for pg_id {0}: {1}".format(pg_id, flow_stats))
+        print("Latency stats for pg_id {0}: {1}".format(pg_id, lat_stats))
+        lat = lat_stats['latency']
+        jitter = lat['jitter']
+        avg = lat['average']
+        tot_max = lat['total_max']
+        tot_min = lat['total_min']
+        last_max = lat['last_max']
+        hist = lat['histogram']
+        total_rx = flow_stats['rx_pkts']['total']
+        total_tx = flow_stats['tx_pkts']['total']
+        drops = lat_stats['err_cntrs']['dropped']
+        ooo = lat_stats['err_cntrs']['out_of_order']
+        dup = lat_stats['err_cntrs']['dup']
+        sth = lat_stats['err_cntrs']['seq_too_high']
+        stl = lat_stats['err_cntrs']['seq_too_low']
+        print("Latency info for pg_id {0}:".format(pg_id))
+        print("  Dropped packets: {0}".format(drops))
+        print("  Maximum latency: {0} us".format(tot_max))
+        print("  Minimum latency: {0} us".format(tot_min))
+        print("  Maximum latency in last sampling period: {0} us".format(last_max))
+        print("  Average latency: {0} us".format(avg))
+        print("  Jitter: {0} us".format(jitter))
+        print("  Latency distribution histogram:")
+        l = list(hist.keys()) # need to listify in order to be able to sort them.
+        l.sort()
+        for sample in l:
+            range_start = sample
+            if range_start == 0:
+                range_end = 10
+            else:
+                range_end  = range_start + pow(10, (len(str(range_start))-1))
+            val = hist[sample]
+            print ("    Packets with latency between {0} us and {1} us: {2}".format(range_start, range_end, val))
+
+        # Get statistics for TX and RX ports
+        for port in ALL_PORTS:
+            readable_stats = get_readable_port_stats(stats[port])
+            print("Statistics for port {}: {}".format(port, readable_stats))
+
+        # Check SLAs.
+        assert(total_rx > 0), "No control traffic has been received"
+        assert(drops == 0), "Control traffic has been dropped: {}".format(drops)
+        assert(sth == 0 and stl == 0), "Control traffic has been delayed or dropped: {}, {}".format(sth, stl)
```

Comment:
I don't understand how `sth` and `stl` can be an indication of delay. To me, it seems that can be used to identify dropped or out-of-order packets, but not delay (unless by delay you mean that an earlier packet is delivered later, hence out-of-order). Am I missing something? I tried to find some docs about `seq_too_high` and `seq_too_low` but couldn't find anything, so it could be I'm misinterpreting their meaning.

Reply:
> For calculating packet error events, we add sequence number to each  packet’s payload. We decide what went wrong only according to sequence  number     of last packet received and that of the previous packet. seq_too_low and seq_too_high count events we see. dup, out_of_order and dropped     are heuristics we apply to try and understand what happened. They will be accurate in common error scenarios.     We describe few scenarios below to help understand this.

https://trex-tgn.cisco.com/trex/doc/trex_stateless.html#_tutorial_simple_interleaving_streams

Reply:
I now see that delay is not applicable here.

Reply:
```suggestion
        assert(sth == 0 and stl == 0), "Control traffic has been dropped or reordered: {}, {}".format(sth, stl)
```